hbase-client/src/main/java/org/apache/hadoop/hbase/client/RawAsyncTableImpl.java

```diff
@@ -796,75 +804,165 @@ private boolean locateFinished(RegionInfo region, byte[] endKey, boolean endKeyI
+          if (e == null && r != null) {
+            updatedCallable = callback.getNextCallable(r, region);
+          } else {
+            updatedCallable = null;
+          }
 
-        // If updatedCallable is non-null, we will be sending another request, so no need to
-        // decrement unfinishedRequest (recall that && short-circuits).
-        // If updatedCallable is null, and unfinishedRequest decrements to 0, we're done with the
-        // requests for this coprocessor call.
-        if (
-          updatedCallable == null && unfinishedRequest.decrementAndGet() == 0
-            && locateFinished.get()
-        ) {
-          callback.onComplete();
-        } else if (updatedCallable != null) {
-          Duration waitInterval = callback.getWaitInterval(r, region);
-          LOG.trace("Coprocessor returned incomplete result. "
-            + "Sleeping for {} before making follow-up request.", waitInterval);
-          if (!waitInterval.isZero()) {
-            AsyncConnectionImpl.RETRY_TIMER.newTimeout(
-              (timeout) -> coprocessorServiceUntilComplete(stubMaker, updatedCallable, callback,
-                locateFinished, unfinishedRequest, region, span),
-              waitInterval.toMillis(), TimeUnit.MILLISECONDS);
+          // If updatedCallable is non-null, we will be sending another request, so no need to
+          // decrement unfinishedRequest (recall that && short-circuits).
+          // If updatedCallable is null, and unfinishedRequest decrements to 0, we're done with the
+          // requests for this coprocessor call.
+          if (updatedCallable == null) {
+            progress.markRegionFinished(region);
           } else {
-            coprocessorServiceUntilComplete(stubMaker, updatedCallable, callback, locateFinished,
-              unfinishedRequest, region, span);
+            Duration waitInterval = callback.getWaitInterval(r, region);
+            LOG.trace("Coprocessor returned incomplete result. "
+              + "Sleeping for {} before making follow-up request.", waitInterval);
+            if (!waitInterval.isZero()) {
+              AsyncConnectionImpl.RETRY_TIMER.newTimeout(
+                (timeout) -> coprocessorServiceUntilComplete(stubMaker, updatedCallable, callback,
+                  progress, region, claimedRegionAssignment, span),
+                waitInterval.toMillis(), TimeUnit.MILLISECONDS);
+            } else {
+              coprocessorServiceUntilComplete(stubMaker, updatedCallable, callback, progress,
+                region, claimedRegionAssignment, span);
+            }
           }
         }
+      } catch (Throwable t) {
+        if (claimedRegionAssignment.get() || progress.tryClaimRegionAssignment(region)) {
+          LOG.error("Error processing coprocessor service response for region {}",
+            region.getEncodedName(), t);
+          progress.onResponse(region, null, t);
+        }
+        progress.markRegionFinished(region);
```

Comment:
I noticed that if the `getNextCallable` throws, we never call `onComplete`; this catch clause fixes that